File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        
        output = [] 
        nums = sorted(nums)
        unique = set() 
        print(nums)
        
        for i in range(0, len(nums) - 2):
            
            first_value = nums[i]
            leftPointer = i + 1 
            rightPointer = len(nums) - 1
            
            while leftPointer < rightPointer: 
                second_value = nums[leftPointer]
                third_value = nums[rightPointer]
                
                if first_value + second_value + third_value == 0:
                    if (first_value, second_value, third_value) not in unique:
                        output.append([first_value, second_value, third_value])
                        unique.add((first_value, second_value, third_value))
                    
                    leftPointer += 1 
                    rightPointer -= 1
                
                elif first_value + second_value + third_value > 0:
                    rightPointer -= 1 
                
                else:
                    leftPointer += 1 
            
        return output 
                    
                
                
            
            
            

        
        
        

File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> output = new ArrayList<>();
        HashSet<List<Integer>> unique = new HashSet<>();
        Arrays.sort(nums);

        for (int i = 0; i < nums.length - 2; i++) {
            int valueOne = nums[i];

            int leftPointer = i + 1;
            int rightPointer = nums.length - 1;

            while (leftPointer < rightPointer) {
                int valueTwo = nums[leftPointer];
                int valueThree = nums[rightPointer];

                if (valueOne + valueTwo + valueThree == 0) {
                    List<Integer> target = new ArrayList<>();
                    target.add(valueOne);
                    target.add(valueTwo);
                    target.add(valueThree);

                    if (!unique.contains(target)) {
                        output.add(target);
                        unique.add(target);
                    }
                    
                    leftPointer += 1; 
                    rightPointer -= 1; 
                    
                } else if (valueOne + valueTwo + valueThree > 0) {
                    rightPointer -= 1;
                } else {
                    leftPointer += 1;
                }
            }
        }
        return output;
    }
}